### PR TITLE
[WIP] Add translation string support

### PR DESCRIPTION
**Pull request by hamishwillee**
@dagar This is my first shot at adding translation string support. Gus still needs to add support for the format described in part 3. This is apparently easy as it is a direct copy of what he does for the camera definition files.

It has three main parts:

1. Generation of parameter_strings.xml containing map of keys to English strings.
  * Created by a new option `--translation_xml` in *_px_process_params.py_
  * This is called in the cmakelists to copy new file to **parameters/translations/parameter_strings.xml**
  * Format looks like below. Has unique key based on param name and subinfo - ie enumvalue, bit index, short/long description. Groups and Categories are "Bare - ie no special key because they have no sub items. We could add group as a parent in parameter index but decided not to.
 ```
 <?xml version='1.0' encoding='UTF-8'?>
 <!DOCTYPE properties SYSTEM "http://java.sun.com/dtd/properties.dtd">
 <properties>
 <entry key="Attitude Q estimator">Attitude Q estimator</entry>
 <entry key="ATT_ACC_COMP_short_desc">Acceleration compensation based on GPS velocity</entry>
 <entry key="ATT_EXT_HDG_M_enumval_1">Vision</entry>
 <entry key="EKF2_AID_MASK_bit_5">multi-rotor drag fusion</entry>
 </properties>
 ```
1. Crowdin integration
  * We can set up crowdin to monitor **Firmware/translations/parameter_strings.xml** in github and spit out translated files with format **Firmware/translations/parameter_strings-xx_YY.xml** (e.g. zh_CH) in the same directory.
1. Addition of localisation to the parameters.xml
  * We add a `localization` to end of `parameters` that has `locale` tags of following format for each locale.
 ```
 ...
 </group>
 <localization>
 <locale name="zh_CN">
 <strings original="Vision" translated="视觉训练" />
 <strings original="Motion Capture" translated="运动捕捉" />
 <strings original="Complimentary filter external heading weight" translated="互补过滤器外部航向权重值" />
 <strings original="Acceleration compensation based on GPS&#10;velocity" translated="基于 GPS 的加速度补偿&#10;速度" />
 <strings original="Complimentary filter magnetometer weight" translated="互补滤波器磁罗盘权重值" />
 <strings original="Attitude Q estimator" translated="姿态 Q 估计器" />
 <strings original="Battery Calibration" translated="电池校准" />
 <strings original="Magnetic declination, in degrees" translated="磁偏角, 以角度为单位" />
 <strings original="Gyro bias limit" translated="陀螺偏置极限值" />
 </locale>
 </localization>
 </parameters>
 ```
  * The code parses Firmware/translations and generates a dictionary of language dictionaries. The translations dict ONLY have translated terms.
  * This is passed into the xmlout to generate the locale information. 

This will not fall over if the translation files are missing. 

Once this (or something like it) is agreed we can add the crowdin integration.

### CODE REVIEW

**Comment by hamishwillee** [comment]
PS I am fully expecting you to hate my Python. That can be tidied if the overall approach is considered OK.

**Comment by bys1123** [comment]
Thanks for your work, we will join translation when it's ready.

**Comment by hamishwillee** [comment]
@dagar The build now passes - the failure appears unrelated to this. 

Can you please ETA review?

File: CMakeLists.txt [hunk 1]
```diff
@@ -484,6 +484,7 @@ file(GLOB_RECURSE yaml_config_files ${PX4_SOURCE_DIR}/src/modules/*.yaml
 	${PX4_SOURCE_DIR}/src/drivers/*.yaml ${PX4_SOURCE_DIR}/src/lib/*.yaml)
 add_custom_target(metadata_parameters
 	COMMAND ${CMAKE_COMMAND} -E make_directory ${PX4_BINARY_DIR}/docs
+        COMMAND ${CMAKE_COMMAND} -E make_directory ${PX4_SOURCE_DIR}/translations
```
Comment (dagar) [inline CMakeLists.txt]: whitespace

**Comment by dagar** [comment]
We should add something simple to CI (Jenkins) that runs this. Next we'd have it post to Crowdin automatically on an overall successful master build.

Example Generate Metadata -
https://github.com/PX4/Firmware/blob/master/Jenkinsfile#L291-L331
Runs on Jenkins here - http://ci.px4.io:8080/blue/organizations/jenkins/PX4%2FFirmware/detail/master/440/pipeline

**Comment by hamishwillee** [comment]
> We should add something simple to CI (Jenkins) that runs this. 

Yes.

> Next we'd have it post to Crowdin automatically on an overall successful master build.

I don't think we want to do this. Instead we want to automatically save **/src/lib/parameters/translations/parameter_strings.xml** into the github tree. Crowdin can be set up to automatically suck out updates and export a PR to spit out the translated results into the same directly.

File: CMakeLists.txt (hunk 1 above)
Comment (hamishwillee) [reply CMakeLists.txt]: Fixed. I added some other changes too to tidy the output.

**Comment by bys1123** [comment]
We tried to run `make parameters_metadata`, looks like few `long_desc` is missing in `parameter_strings.xml`. 
eg. WEST_BETA_GATE, WEST_TAS_GATE
Is this already clean up duplicated strings?
@hamishwillee

File: src/lib/parameters/px4params/xmlstringsout.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+class XMLOutput():
+
+    def __init__(self, groups, board, inject_xml_file_name):
+        unique_strings = set()
+        xml_parameters = ET.Element("properties")
+
+        for group in groups:
+            group_name=group.GetName()
+            if group_name not in unique_strings:
```
Comment (bys1123) [inline src/lib/parameters/px4params/xmlstringsout.py]:
```py
            if group_name not in unique_strings:
```
Oh, looks like it is

Comment (hamishwillee) [reply src/lib/parameters/px4params/xmlstringsout.py]: Indeed :-)

This is waiting on @dagar - he will review and merge (if appropriate) when it gets near the top of his priority list.

**Comment by bys1123** [comment]
(っ•̀ω•́)っ✎⁾⁾

**Comment by hamishwillee** [comment]
@bkueng FYI, this is languishing, but the problem of parameter translation is worth thinking about when you look at the broader "strings in PX4" translation support.

**Comment by bys1123** [comment]
@hamishwillee Hey, What is the stage of this PR now?

**Comment by hamishwillee** [comment]
@bys1123 Nowhere.
I think that this will be addressed in part by @bkueng work in the events interface. 
But this PR is now so out of sync we'd have to start again to merge it. I'm therefore closing.

**Pull request closed without merging**